src/lib.rs

```diff
@@ -846,6 +856,17 @@ impl<'a> ParserState<'a> {
             if self.consume(b"$A6") {
                 return self.read_func_type();
             }
+            if let Some(x) = self.peek() {
+                match x {
+                    // Inheritance specifiers, which we don't need to remember.
+                    b'1' | b'H' | b'I' | b'J' => {
```

Comment:
This is not quite right.  I think we should only be checking for this in certain contexts, and the current checking is not strigent enough.

Reply:
e.g. prior to this commit, we can demangle `??$emplace_hint@AEBUpiecewise_construct_t@std@@V?$tuple@AEBH@2@V?$tuple@$$V@2@@?$_Tree@V?$_Tmap_traits@HUPayload@RtpUtility@webrtc@@U?$less@H@std@@V?$allocator@U?$pair@$$CBHUPayload@RtpUtility@webrtc@@@std@@@5@$0A@@std@@@std@@QEAA?AV?$_Tree_iterator@V?$_Tree_val@U?$_Tree_simple_types@U?$pair@$$CBHUPayload@RtpUtility@webrtc@@@std@@@std@@@std@@@1@V?$_Tree_const_iterator@V?$_Tree_val@U?$_Tree_simple_types@U?$pair@$$CBHUPayload@RtpUtility@webrtc@@@std@@@std@@@std@@@1@AEBUpiecewise_construct_t@1@$$QEAV?$tuple@AEBH@1@$$QEAV?$tuple@$$V@1@@Z` without issues, but such a symbol fails to demangle after this commit.

Reply:
We should split read_var_type() into read_template_param_type(). It can handle the template specific situations like b'1' | b'H' | b'I' | b'J'  and then fall back to the remaining read_var_type() as needed.